File: lb3/java/Producer.java
public class Producer extends Thread {
    private final Storage _store;
    private final int _consumes;
    private final int _id;

    public Producer(int consumes, Storage store, int id) {
        this._store = store;
        this._consumes = consumes;
        this._id = id;
    }

    @Override
    public void run() {
        for (int i = 0; i < this._consumes; i++) {
            this._store.waitForDiminution();
            ThreadSafeLogger.log("Producer " + this._id + " waited for diminution");

            this._store.enter();
            ThreadSafeLogger.log("Producer " + this._id + " entered");

            var item = this.getProductName("" + i);
            this._store.addItem(item);
            ThreadSafeLogger.log("Producer " + this._id + " added item: " + item);

            this._store.leave();
            ThreadSafeLogger.log("Producer " + this._id + " left");
        }
    }

    private String getProductName(String label) {
        return "producer-" + this._id + "-" + "product-" + label;
    }
}

File: lb3/c#/lb3c#/Producer.cs
namespace lb3;

class Producer : Operator {
    public Producer(int produces, Storage store, int id) : base(produces, store, id) {}

    protected override void Operate() {
        for (int i = 0; i < this._operations; i++) {
            this._store.WaitForDiminution();
            System.Console.WriteLine("Producer " + this._id + " waited for diminution");
            
            this._store.Enter();
            System.Console.WriteLine("Producer " + this._id + " entered");
            
            var item = this.GetProductName($"{i}");
            this._store.AddItem(item);
            System.Console.WriteLine("Producer " + this._id + " added item: " + item);

            this._store.Leave();
            System.Console.WriteLine("Producer " + this._id + " left");
        }
    }

    private string GetProductName(string label) {
        return $"Producer-${this._id}-product-{label}";
    }
}
